File: system/webui/database/transaction.py
"""Transaction"""

from typing import Any, Literal

from sqlite_database import Row
from sqlite_database.operators import le, ge
from ...database import transaction, generate_id, database


class TransactionAPI:
    """Transaction"""

    def _validate(self, data: dict[str, Any]):
        """Validate transaction data"""
        required_fields = ["type", "category_id", "amount", "date"]
        for field in required_fields:
            if field not in data:
                raise ValueError(f"Missing required field: {field}")

        if not isinstance(data["type"], str):
            raise TypeError("Type must be a string")

        if not isinstance(data["category_id"], str):
            raise TypeError("Category ID must be a string")

        if not isinstance(data["amount"], (int, float)):
            raise TypeError("Amount must be a number")

        if not isinstance(data['time'], float):
            raise TypeError("Time must be a float")

        if "notes" in data and not isinstance(data["notes"], str):
            raise TypeError("Notes must be a string")

    def add_transaction(self, data: dict[str, Any]):
        """Add a transaction to the database"""
        self._validate(data)
        transaction.insert(
            {
                "id": generate_id(),
                "type": data["type"],
                "category_id": data["category_id"],
                "amount": data["amount"],
                "time": data["time"],
                "notes": data.get("notes", "<no note>"),
            }
        )

    def edit_transaction(self, data: dict[str, Any]):
        """Edit a transaction in the database"""
        self._validate(data)
        transaction.update(
            data["id"],
            {
                "type": data["type"],
                "category_id": data["category_id"],
                "amount": data["amount"],
                "time": data["time"],
                "notes": data.get("notes", "<no note>"),
            },
        )

    def delete_transaction(self, transaction_id: str):
        """Delete a transaction from the database"""
        transaction.delete(transaction_id)

    def get_transaction(self, transaction_id: str):
        """Get a transaction from the database"""
        return transaction.select_one({"id": transaction_id})

    def get_transactions(self, limit=0):
        """Get all transactions from the database"""
        return transaction.select(limit=limit)

    def get_transactions_by_category(self, category_id: str, limit=0):
        """Get all transactions for a given category"""
        return transaction.select({"category_id": category_id}, limit=limit)

    def get_transactions_by_time(self, date: float, state: Literal["newer", "older"], limit=0):
        """Get all transactions for a given time"""
        return transaction.select(
            [le("time", date) if state == "older" else ge("time", date)],
            limit=limit
        )

    def get_transactions_by_type(self, tr_type: str, limit=0) -> list[Row]:
        """Get all transactions for a given type"""
        return transaction.select({"type": tr_type}, limit=limit)

    def get_transactions_by_amount(self, amount: float, state: Literal["greater", "less"], limit):
        """Get all transactions for a given amount"""
        return transaction.select(
            [ge("amount", amount) if state == "greater" else le("amount", amount)],
            limit=limit
        )

    def get_ungrouped_transactions(self) -> list[Row]:
        """Get all ungrouped transactions"""
        query = '''
                SELECT t.*
                FROM transaction_tbl t
                LEFT JOIN collection_transaction ct ON t.id = ct.transaction_id
                WHERE ct.transaction_id IS NULL
            '''
        return database.sql.execute(query).fetchall()

File: data/static/js/models/transaction.mjs
import {system} from "../init.mjs"
/** @typedef {import("../types.mjs").Transaction} BaseTransaction */

class Transaction {

  /**
   * Transaction class
   * @param {string} id 
   * @param {string} type 
   * @param {string} category 
   * @param {number} amount 
   * @param {number} time 
   * @param {string} notes 
   */
  constructor(id, type, category, amount, time, notes) {
    this.id = id;
    this.type = type;
    this.category = category
    this.amount = amount
    this.time = time;
    this.notes = notes
  }

  /**
   * Push current transaction to database
   */
  async push() {
      system.db.transaction.add_transaction(this)
  }

  /**
   * Update current transaction
   */
  async update() {
    system.db.transaction.update_transaction(this)
  }

  /**
   * Delete current transaction
   */
  async delete() {
    system.db.transaction.delete_transaction(this.id)
  }

  /**
   * cast base transaction to array of transaction
   * @param {BaseTransaction[]} arrayBase 
   * @param {Transaction[]} target 
   */
  static #castTo(arrayBase, target) {
    arrayBase.forEach(e => {
      target.push(new Transaction(e.id, e.type, e.category_id, e.amount, e.time, e.notes))
    });
  }

  /**
   * Get transaction from database
   * @param {string} id 
   * @returns {Transaction}
   */
  static async get(id) {
    /** @type {BaseTransaction} */ const base = system.db.transaction.get_transaction(id)
    return new Transaction(base.id, base.notes, base.type, base.category_id, base.amount, base.time, base.notes)
  }

  /**
   * Get all ungrouped transaction from database
   * @returns {Promise<Transaction[]>}
   */
  static async get_ungrouped() {
    const r = []
    const base = await (system.db.transaction.get_ungrouped_transactions())
    this.#castTo(base, r)
    return r
  }

  /**
   * Get all transaction from database
   * @param {number} [limit=0] 
   * @param {string} [order=null] order of operation. Either asc/desc
   * @returns {Promise<Transaction[]>}
   */
  static async get_all(limit = 0, order = null) {
    const r = []
    const base = await system.db.transaction.get_transactions(limit, order)
    this.#castTo(base, r)
    return r
  }

  /**
   * Get transactions based on amount
   * @param {number} amount 
   * @param {string} state whether filtering less by amount or greater than certain amount (key: greater/lesser)
   * @param {number} limit 
   * @param {string} [order=null] order of operation. Either asc/desc
   * @returns {Promise<Transaction[]>}
   */
  static async get_by_amount(amount, state, limit = 0, order=null) {
    /** @type {Transaction[]} */ const r = []
    const base = await system.db.transaction.get_transaction_by_amount(amount, state, limit, order)
    this.#castTo(base, r)
    return r
  }

  /**
   * Get transactions based on category
   * @param {string} cid 
   * @param {number} limit 
   * @param {string} [order=null] order of operation. Either asc/desc
   * @returns {Promise<Transaction[]>}
   */
  static async get_by_category(cid, limit = 0, order=null) {
    const r = []
    const base = await system.db.transaction.get_transactions_by_category(cid, limit, order)
    this.#castTo(base, r)
    return r
  }

  /**
   * Get transactions based on time/date
   * @param {number} date 
   * @param {string} state whether filtering before (older) or after (newer) (key: older/newer)
   * @param {number} limit 
   * @param {string} [order=null] order of operation. Either asc/desc
   * @returns {Promise<Transaction[]>}
   */
  static async get_by_time(date, state, limit = 0, order=null) {
    const r = []
    const base = await system.db.transaction.get_transactions_by_time(date, state, limit, order)
    this.#castTo(base, r)
    return r
  }

  /**
   * Get transactions based on type
   * @param {number} tr_type 
   * @param {number} limit 
   * @param {string} [order=null] order of operation. Either asc/desc
   * @returns {Promise<Transaction[]>}
   */
  static async get_by_type(tr_type, limit = 0, order=null) {
    const r = []
    const base = await system.db.transaction.get_transactions_by_type(tr_type, limit, order)
    this.#castTo(base, r)
    return r
  }
}

export { Transaction }